### ISSUE TITLE
False positive for the rule CA1001 when creating an instance of an IDisposable type in a struct and assigning it to a field.

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn-analyzers#7095]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/False-positive-for-the-rule-CA1001-when-/10539466)._

---
\[severity:It bothers me. A fix would be nice\]  
In the following code, Visual Studio reports CA1001 (“Types that own disposable fields should be disposable”) on the struct `MyCursor` even if it actually implements `IDisposable`.

```  
using System;
using System.Collections;
using System.Collections.Generic;

namespace Issue20231212 {
	public class Foo : IDisposable {
		protected virtual void Dispose(bool disposing) { }
		public void Dispose() {
			Dispose(true);
			GC.SuppressFinalize(this);
		}
	}
	internal class FooWrapper(Foo foo) : IDisposable {
		readonly Foo _foo = foo;
		protected virtual void Dispose(bool disposing) {
			if (disposing) _foo.Dispose();
		}
		public void Dispose() {
			Dispose(true);
			GC.SuppressFinalize(this);
		}
	}
	internal readonly struct MyCursor : IEnumerator<object?>, IEnumerator, IDisposable {
		readonly FooWrapper _wrapper;
		internal MyCursor(Foo foo) {
			_wrapper = new FooWrapper(foo);
		}
		public object? Current => null;
		public void Dispose() { _wrapper.Dispose(); }
		public bool MoveNext() { throw new NotImplementedException(); }
		public void Reset() { throw new NotImplementedException(); }
	}
}
```

![20231212230644.png](https://aka.ms/dc/image?name=B196feb509f4d4656ab61d85ff5425c3e638379904349539879_20231212230644.png&tid=196feb509f4d4656ab61d85ff5425c3e638379904349539879)

---
#### Feedback Bot on 12/12/2023, 08:25 PM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

**Comment by sharwell** [comment]
Duplicate of #6151
